Introduce withSearchedProducts higher order component and refactor ProductsControl Component

nerrad opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

As outlined in p90Yrv-1an-p2, this pull refactors the `ProductsControl` component so that data/logic is kept separate from presentation.  So with this pull:

- introduction of new `withSearchedProducts` higher order component handling the logic related to getting products (and updating `onSearch`).
- adds unit tests for the new hoc.
- refactoring `ProductsControl` component so that it now is solely presentation and receives data via the hoc.

### How to test the changes in this Pull Request:

This should have no impact to behaviour of the block for end users as this is solely a internal change.  With that said, the following covers tests:

* [ ] Unit tests should pass (including the new tests for the hoc).
* [ ] Existing functionality of the "Hand Picked Products" block should continue to work as expected (specifically make sure that the search functionality works and multiple blocks in the same view work with different selected products).
* [ ] Verify there are no console errors introduced by work in this branch.

### Changelog

> Internal: Introduce new `withSearchedProducts` higher order component and refactor `ProductsControl` component to use it.

Aljullu commented on assets/js/components/products-control/index.js:
```diff
@@ -2,110 +2,91 @@
-	render() {
-		const { list, loading } = this.state;
-		const { onChange, selected } = this.props;
+import { withSearchedProducts } from '../../hocs/with-searched-products';
```
You can import it directly from `index.js`, right?
```suggestion
import { withSearchedProducts } from '../../hocs';
```

nerrad replied:
Yup, good catch!  Fixed in 34cd8ed (another place I could've just committed your suggestion).

Aljullu commented on assets/js/hocs/index.js:
```diff
@@ -0,0 +1 @@
+export { default as withSeachedProducts } from './with-searched-products';
```
Small typo here:
```suggestion
export { default as withSearchedProducts } from './with-searched-products';
```

Aljullu commented on assets/js/components/products-control/index.js:
```diff
@@ -2,110 +2,91 @@
+	onSearch: PropTypes.func,
+	selected: PropTypes.array,
+	products: PropTypes.array,
+	isLoading: PropTypes.boolean,
```
It should be `bool` here:
```suggestion
	isLoading: PropTypes.bool,
```

nerrad commented on assets/js/hocs/test/with-searched-products.js:
```diff
@@ -0,0 +1,83 @@
+			expect( props.products ).toEqual( selected );
+		} );
+		it( 'debounce and getProducts is called on search event', () => {
+			props.onSearch();
```
See fix in 8e7533d

I decided to just repeat the props retrieval like you suggested here (I could've just accepted this suggestion but uh, got ahead of myself 😛

nerrad commented:
> Can you reproduce it too? I also added some minor comments below.

Doh!  I missed an important step in the setup and forgot to run `composer install` for the plugin.  Thus I was testing against what is bundled in Woo core vs the plugin build for this branch 🤦‍♂ .  Once I addressed that, I retested and reproduced what you saw and from that discovered the mistake I made in the refactor in da6f732

I've noted with your other comments the commit fixing things.

Aljullu commented on assets/js/components/products-control/index.js:
```diff
@@ -2,110 +2,91 @@
+
+ProductsControl.defaultProps = {
+	onChange: () => undefined,
+	onSearch: () => undefined,
```
I don't know if there is any performance difference, but couldn't this be `null` instead of `() => undefined`? We already [check for `onSearch` type](https://github.com/woocommerce/woocommerce-admin/blob/master/packages/components/src/search-list-control/index.js#L53).

It shouldn't make any difference now, but at some point `<SearchListControl>`'s implementation might change and it might set the button to `disabled` if `onSearch` is falsy or something like that. In that case passing a function like this would bring unexpected results.

nerrad replied:
In the `<SearchListControl>` component is this:

https://github.com/woocommerce/woocommerce-admin/blob/ca66af481684642cae88da79706d33b978c9bf49/packages/components/src/search-list-control/index.js#L290-L298

`onChange` is listed as requiring a function and `onSearch` is optional.  `<ProductsControl>` is essentially a wrapper around `<SearchListControl>` (which get's enhanced by `withSearchedProducts`).  So the question becomes whether we want to enforce the requirements as the wrapped component (and would probably want to keep in sync with whatever changes happen in the SearchListControl) or we don't enforce any checks on this component and allow the flow to go through to SearchListControl to throw any errors.

I think the former option would be a bit preferable so I'll implement a required check on `onChange` for ProductsControl . I believe it's preferable because it's better to fail as early as possible for consuming code to have an awareness of what needs fixed (i.e. point to incorrect usage of `<ProductsControl>` rather than `<SearchListControl>` which in this case is a hidden implementation.  I'll make that change.

As for _why_ I implemented things this way in the first place - I was thinking there might be usecases where consuming code might simply not pass in any callbacks on `onChange` or `onSearch` and thus this would allow for implementing the component without them.  But in hindsight that seems like a very edgecase (and non-useful) implementation :).

nerrad replied:
changed in a3f49ce

Aljullu commented on assets/js/hocs/with-searched-products.js:
```diff
@@ -0,0 +1,86 @@
+					isLoading={ loading }
+					selected={ selected
+						.map( ( id ) => find( list, { id } ) )
+						.filter( Boolean ) }
```
I know that's not new from this PR either, but if I'm not missing any edge case this can be rewritten and avoid the lodash dependency:
```ES
selected={ list.filter( ( { id } ) => selected.includes( id ) ) }
```

nerrad replied:
fixed in e48b140

nerrad commented on assets/js/components/products-control/index.js:
```diff
@@ -2,110 +2,91 @@
+		),
+	};
+	return (
+		<Fragment>
```
nope

nerrad replied:
changed in d8ea0af
